### PR TITLE
XA support for Galera

### PR DESCRIPTION
**Pull request by sciascid**
I'm opening this PR to get early feedback/review, although notice that it should probably rebased on top of https://github.com/MariaDB/server/pull/1183

### CODE REVIEW

File: sql/wsrep_client_service.cc [hunk 1]
```diff
@@ -154,6 +154,13 @@ bool Wsrep_client_service::is_xa() const
   return m_thd->transaction.xid_state.xa_state != XA_NOTR;
 }
 
+bool Wsrep_client_service::is_xa_prepare() const
+{
+  DBUG_ASSERT(m_thd == current_thd);
+  THD* thd= m_thd;
```
Comment (janlindstrom) [inline sql/wsrep_client_service.cc]: Why you need this extra variable ?

**Comment by sciascid** [comment]
I squashed the commits into one commit, and rewrote the commit message, mentioning MDEV-17099

File: sql/log.cc [hunk 2]
```diff
@@ -5687,20 +5687,39 @@ THD::binlog_start_trans_and_stmt()
...
+
+      if (this->transaction.xid_state.xa_state == XA_ACTIVE)
+      {
+        Format_description_log_event fd_ev(4);
+        fd_ev.checksum_alg= BINLOG_CHECKSUM_ALG_OFF; //(enum_binlog_checksum_alg)binlog_checksum_options;
+
+        String xa_start;
+        xa_start.append(STRING_WITH_LEN("XA START "));
+        wsrep_append_xid_string(xa_start, &this->transaction.xid_state.xid);
+        Query_log_event qinfo(this, "XA START 'test'", strlen("XA START 'test'"),
```
[the remark this answers is not in the document]
Comment (sciascid) [reply sql/log.cc]: This has been fixed.

File: sql/wsrep_applier.cc [hunk 3]
```diff
@@ -176,6 +179,18 @@ int wsrep_apply_events(THD*        thd,
       }
     }
 #endif /* GTID_SUPPORT */
+    case QUERY_EVENT:
+      WSREP_DEBUG("query event: %s len: %d", ((Query_log_event*)ev)->query, ((Query_log_event*)ev)->q_len);
+      // Ugly hack coming next...
```
[the remark this answers is not in the document]
Comment (sciascid) [reply sql/wsrep_applier.cc]: The comment has been rewritten, let me know if that clarifies

File: sql/wsrep_client_service.cc (hunk 1 above)
Comment (sciascid) [reply sql/wsrep_client_service.cc]: Looks unnecessary. Will remove it

**Comment by sciascid** [comment]
@andrelkin fixed

**Comment by sciascid** [comment]
Closing this pull request. Will further work on it and resubmit at a later time.

**Pull request closed without merging**